### PR TITLE
refactor: Fix link error with --enable-debug

### PR DESCRIPTION
**Pull request by hebasto**
Fixes a link error on master (39bd9ddb8783807b9cde6288233e86ad7c85d61f):
```
$ ./configure --enable-debug
$ make
...
bitcoin_wallet-bitcoin-wallet.o:(.data.rel.ro+0x0): undefined reference to `InitError(bilingual_str const&)'
libbitcoin_wallet_tool.a(libbitcoin_wallet_tool_a-wallettool.o):(.data.rel.ro+0x8): undefined reference to `InitError(bilingual_str const&)'
libbitcoin_wallet.a(libbitcoin_wallet_a-salvage.o):(.data.rel.ro+0x8): undefined reference to `InitError(bilingual_str const&)'
libbitcoin_wallet.a(libbitcoin_wallet_a-wallet.o):(.data.rel.ro+0x8): undefined reference to `InitError(bilingual_str const&)'
libbitcoin_wallet.a(libbitcoin_wallet_a-walletdb.o):(.data.rel.ro+0x8): undefined reference to `InitError(bilingual_str const&)'
libbitcoin_wallet.a(libbitcoin_wallet_a-wallet.o):(.data.rel.ro+0x8): more undefined references to `InitError(bilingual_str const&)' follow
collect2: error: ld returned 1 exit status
```

See:
- https://github.com/bitcoin/bitcoin/pull/19295#issuecomment-645471771
- https://github.com/bitcoin/bitcoin/pull/19295#issuecomment-645487182

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Why does the given patch fix this?

**Comment by achow101** [comment]
ACK ef11f39c7371898b99182ea92899e2c9698d7391

Tested that I can compile with `--enable-debug`. Also no idea why this works.

**Comment by Sjors** [comment]
Could use some documentation as to why ef11f39 fixes the issue, but it does.

**Comment by hebasto** [comment]
> How does the given patch fix this?

Not sure ((

It seems related to instantiating of the `FatalError()` template function, and all used symbols should be in the same translation unit. But don't know how this depends on `--enable-debug`.

**Comment by hebasto** [comment]
FWIW, I've managed to reproduce the initial issue with the only flag:
```
$ ./configure CXXFLAGS=-O0
```

**Comment by hebasto** [comment]
The alternative approach is not using function aliasing at all:

```diff
diff --git a/src/ui_interface.cpp b/src/ui_interface.cpp
index 15795bd67..c6017e57a 100644
--- a/src/ui_interface.cpp
+++ b/src/ui_interface.cpp
@@ -59,6 +59,11 @@ bool InitError(const bilingual_str& str)
     return false;
 }
 
+bool AbortError(const bilingual_str& str)
+{
+    return InitError(str);
+}
+
 void InitWarning(const bilingual_str& str)
 {
     uiInterface.ThreadSafeMessageBox(str, "", CClientUIInterface::MSG_WARNING);
diff --git a/src/ui_interface.h b/src/ui_interface.h
index 356d30eaf..bfa771a1b 100644
--- a/src/ui_interface.h
+++ b/src/ui_interface.h
@@ -122,7 +122,7 @@ void InitWarning(const bilingual_str& str);
 
 /** Show error message **/
 bool InitError(const bilingual_str& str);
-constexpr auto AbortError = InitError;
+bool AbortError(const bilingual_str& str);
 
 extern CClientUIInterface uiInterface;
 
```

File: src/ui_interface.h [hunk 1]
```diff
@@ -122,7 +122,6 @@ void InitWarning(const bilingual_str& str);
 
 /** Show error message **/
 bool InitError(const bilingual_str& str);
-constexpr auto AbortError = InitError;
```
Comment (MarcoFalke) [inline src/ui_interface.h:125]:
```suggestion
inline bool AbortError(const bilingual_str& str) { return InitError(str); }

```

**Review by maflcko — COMMENTED** [review]
A smaller patch would be to write the verbose version of the alias?

File: src/ui_interface.h (hunk 1 above)
Comment (Sjors) [reply src/ui_interface.h:125]: Which is more readable too.

**Comment by hebasto** [comment]
Updated ef11f39c7371898b99182ea92899e2c9698d7391 -> b83cc0fc94df99f0334430e63e8c9fa6ae3790e1 ([pr19309.01](https://github.com/hebasto/bitcoin/commits/pr19309.01) -> [pr19309.02](https://github.com/hebasto/bitcoin/commits/pr19309.02), [diff](https://github.com/hebasto/bitcoin/compare/pr19309.01..pr19309.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/19309#pullrequestreview-432676615):

> A smaller patch would be to write the verbose version of the alias?

File: src/ui_interface.h (hunk 1 above)
Comment (hebasto) [reply src/ui_interface.h:126]: [Updated](https://github.com/bitcoin/bitcoin/pull/19309#issuecomment-645582741).

**Comment by dongcarl** [comment]
Is there an upstream GCC bug for this? If we decide to go with this workaround, perhaps leave a comment with some context and we can revert in the future?

**Comment by MarcoFalke** [comment]
@dongcarl I don't think anyone has reduced the test case yet

**Comment by achow101** [comment]
Re-ACK b83cc0fc94df99f0334430e63e8c9fa6ae3790e1

**Comment by MarcoFalke** [comment]
This is not a gcc bug, but a Bitcoin Core bug. Proper fix is here: #19331

**Comment by hebasto** [comment]
@MarcoFalke 
> This is not a gcc bug, but a Bitcoin Core bug. Proper fix is here: #19331

Thank you Marco!

**Pull request merged**